PBSPro/pbspro
# PP-1258: Add a M4 macro to enable online data compression in TPP

[PP-1258-PTL-LOG.txt](https://github.com/PBSPro/pbspro/files/2231785/PP-1258-PTL-LOG.txt)




#### Bug/feature Description
The fix enables data compression in TPP layer, if the sizeof TPP data packet is greater than 8192 bytes.
Compression is triggered for jobs with huge attribute data, possibly a huge environment attached to the job.
Data compression has been enabled using libz library by adding M4 macro file.

Compression routines(tpp_deflate and tpp_inflate) are triggered with large tpp data packets with size more than 8192 bytes.
This requires job with large attribute data.
I have used a large environment data file to force tpp to compress and decompress the data.
Used sample env file ( "compress" size > 8192 bytes ) to submit the job as
                `qsub "-vTEST=`cat compress`" -- /bin/sleep 10`
Following messages are logged in Server and Mom log files 
                Server Log:
                `07/04/2018 22:49:34;0d80;Server@pravha;TPP;tpp_send(Main Thread);tpp_deflate called inside tpp_send routine`
               Mom Log:
               `07/04/2018 22:49:34;0d80;pbs_mom;TPP;add_part_packet(Thread 0);tpp_inflate called inside add_part_packet routine`

Compression routines(tpp_deflate and tpp_inflate) are triggered with large tpp data packets with size more than 8192 [bytes.]

GDB Way:
To check tpp_deflate()
1. Attach Server daemon to GDB
2. set break point at tpp_client.c:1044
3. Continue
Submit the job with large attribute data.

To check tpp_inflate()
1. Attach MoM daemon to GDB
2. set break point at tpp_client.c:3555
3. Continue
Submit the job with large attribute data.





#### Affected Platform(s)
Linux

#### Cause / Analysis / Design


#### Solution Description
I have introduced an m4 file [ **m4/with_libz.m4** ] which checks the libz library across the platforms and enables the macro **PBS_COMPRESSION_ENABLED** during the configure stage.
Appropriate Makefile.am files have been updated to link against libz library.

#### Testing logs/output
Introduced a PTL script to verify the fix[ pbs_compress.py ] and attached the output log of the PTL script.
Attached 2 log files server.log and mom.log

#### Checklist:



- [x
[mom.log](https://github.com/PBSPro/pbspro/files/2169116/mom.log)
[server.log](https://github.com/PBSPro/pbspro/files/2169117/server.log)

] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__
